datafusion/core/src/physical_optimizer/projection_pushdown.rs

```diff
@@ -0,0 +1,2181 @@
+pub fn remove_unnecessary_projections(
+    plan: Arc<dyn ExecutionPlan>,
+) -> Result<Transformed<Arc<dyn ExecutionPlan>>> {
+    let maybe_modified = if let Some(projection) =
+        plan.as_any().downcast_ref::<ProjectionExec>()
+    {
+        // If the projection does not cause any change on the input, we can
+        // safely remove it:
+        if is_projection_removable(projection) {
+            return Ok(Transformed::Yes(projection.input().clone()));
+        }
+        // If it does, check if we can push it under its child(ren):
+        let input = projection.input().as_any();
+        if let Some(csv) = input.downcast_ref::<CsvExec>() {
+            try_swapping_with_csv(projection, csv)
+        } else if let Some(memory) = input.downcast_ref::<MemoryExec>() {
+            try_swapping_with_memory(projection, memory)?
+        } else if let Some(child_projection) = input.downcast_ref::<ProjectionExec>() {
+            let maybe_unified = try_unifying_projections(projection, child_projection)?;
+            return if let Some(new_plan) = maybe_unified {
+                // To unify 3 or more sequential projections:
+                remove_unnecessary_projections(new_plan)
+            } else {
+                Ok(Transformed::No(plan))
+            };
+        } else if let Some(output_req) = input.downcast_ref::<OutputRequirementExec>() {
+            try_swapping_with_output_req(projection, output_req)?
+        } else if input.is::<CoalescePartitionsExec>() {
+            try_swapping_with_coalesce_partitions(projection)?
+        } else if let Some(filter) = input.downcast_ref::<FilterExec>() {
+            try_swapping_with_filter(projection, filter)?
+        } else if let Some(repartition) = input.downcast_ref::<RepartitionExec>() {
+            try_swapping_with_repartition(projection, repartition)?
+        } else if let Some(sort) = input.downcast_ref::<SortExec>() {
+            try_swapping_with_sort(projection, sort)?
+        } else if let Some(spm) = input.downcast_ref::<SortPreservingMergeExec>() {
+            try_swapping_with_sort_preserving_merge(projection, spm)?
+        } else if let Some(union) = input.downcast_ref::<UnionExec>() {
+            try_pushdown_through_union(projection, union)?
+        } else if let Some(hash_join) = input.downcast_ref::<HashJoinExec>() {
+            try_pushdown_through_hash_join(projection, hash_join)?
+        } else if let Some(cross_join) = input.downcast_ref::<CrossJoinExec>() {
+            try_swapping_with_cross_join(projection, cross_join)?
+        } else if let Some(nl_join) = input.downcast_ref::<NestedLoopJoinExec>() {
+            try_swapping_with_nested_loop_join(projection, nl_join)?
+        } else if let Some(sm_join) = input.downcast_ref::<SortMergeJoinExec>() {
+            try_swapping_with_sort_merge_join(projection, sm_join)?
+        } else if let Some(sym_join) = input.downcast_ref::<SymmetricHashJoinExec>() {
```

Comment:
could we make this registry-based so that custom execs could also profit from this pass?

Reply:
Perhaps we can make a method on `ExecutionPlan` and then add the relevant methods to each `impl ExecutionPlan`, similar to what I did in https://github.com/apache/arrow-datafusion/pull/7936

Reply:
Maybe we could add a function like this to ExecutionPlan:

```
trait ExecutionPlan {
...
  /// Tries to push a projection of the output of this ExecutionPlan 
  /// *into* itself input by rewriting the internal expressions.
  /// 
  /// For example,  
  /// (TODO EXAMPLE HERE
  ///
  /// If the ExecutionPlan does not support pushdown, ,returns Ok(None) (the default)
  fn push_projection(&self, projection: &[(Arc<dyn PhysicalExpr>, String)]) -> Result<Option<Arc<dyn ExecutionPlan>>>) {
    Ok(None)
  }
}
```

Reply:
So I guess this whole problem is:

> How is a set of optimizer passes linked to a set of nodes, while both sets are extendible?

I see the following rough solutions:

# A: omniscient optimizer

The optimizer knows all node types. This is what this PR does (and what many other passes do).

This doesn't scale well.

# B: omniscient nodes
The nodes know all optimizer passes and implement them themselves. This kinda sounds like what @alamb proposes.

This doesn't scale well.

# C: registry-based linking
Developers are aware of the which nodes can be optimized in which way and can fill out gaps in the optimizer-node matrix. The issue is mostly how this registry should be implemented. Rust has a bunch of crates for that that are all not great (due to the issue of the initialization order). Luckily we all know which node types are in a plan (because you could traverse the plan) so we could hook registry initialization in there. Something like:

```rust
trait ExecutionPlan {
    fn register_hook_for_optimizer_pass(....);
}
```

# D: abstraction
This is what most other optimizer passes do: they read some abstract property of the node (like "schema", "num children", "output ordering", ...) and infer the correct behavior based on that. I think we could use something like this here as well. Namely if you would know what columns of an input schema are used by the node itself and which are just "pass-through", you could apply projection pushdown automatically.

Reply:
IIUC what @alamb proposes is almost at the boundary of categories B and D. If we can define what the proposed function does purely in terms of projection behavior (which would still have meaning independent from the pushdown rule), we can consider it to be in category D.

We think that category A is not in-line with Datafusion's philosophy, and I think we all agree on this. However, in many cases, category A-type implementations serve as a good stepping stone as we try to gain a good understanding of how a good category C/D design looks like. So, on our end, we typically employ the strategy of getting a good test suite and a baseline category A implementation done first, and then progressively migrate towards a long-term category C/D solution. This PR is one of such first steps 🙂

Reply:
I think with a follow-up ticket "make this rule generic" we could accept the solution in this PR, WDYT @alamb ?

Reply:
> I think with a follow-up ticket "make this rule generic" we could accept the solution in this PR, WDYT @alamb ?

I agree -- I filed https://github.com/apache/arrow-datafusion/issues/8096